File: bindings/dotnet/src/Zlink/Contracts/Monitoring/Poller.cs
// SPDX-License-Identifier: MPL-2.0

using System;

namespace Systems.Zlink;

public interface IPoller : IDisposable, IAsyncDisposable
{
    int Size { get; }

    void Add(IZlinkSocket socket, PollEventFlags events, nuint slot);

    void AddFd(int fd, PollEventFlags events, nuint slot);

    void Add(IZlinkTimer timer, nuint slot);

    void Modify(IZlinkSocket socket, PollEventFlags events);

    void ModifyFd(int fd, PollEventFlags events);

    bool Remove(IZlinkSocket socket);

    bool Remove(IZlinkTimer timer);

    bool Remove(int fd);

    void Clear();

    void Close();

    int Wait(Span<PollEvent> destination, TimeSpan timeout);
}

File: bindings/java/src/main/java/systems/zlink/contracts/eventing/Poller.java
/* SPDX-License-Identifier: MPL-2.0 */

package systems.zlink.contracts.eventing;

import systems.zlink.contracts.service.spot.Spot;
import systems.zlink.contracts.sockets.Socket;
import java.time.Duration;

public interface Poller extends AutoCloseable {

    void add(Socket socket, long slot, PollEventFlags... events);

    void add(Spot spot, long slot, PollEventFlags... events);

    void addFd(int fd, long slot, PollEventFlags... events);

    void add(ZlinkTimer timer, long slot);

    void modify(Socket socket, PollEventFlags... events);

    void modify(Spot spot, PollEventFlags... events);

    void modifyFd(int fd, PollEventFlags... events);

    boolean remove(Socket socket);

    boolean remove(Spot spot);

    boolean remove(int fd);

    boolean remove(ZlinkTimer timer);

    void clear();

    int size();

    int wait(PollEvents events, Duration timeout);

    @Override
    void close();
}
